crates/edr_provider/src/data.rs

```diff
@@ -1767,25 +1767,41 @@ impl<LoggerErrorT: Debug> ProviderData<LoggerErrorT> {
             &Box<dyn SyncState<StateError>>,
         ) -> T,
     ) -> Result<T, ProviderError<LoggerErrorT>> {
-        let maybe_block_number = self
-            .block_number_by_block_spec(block_spec.unwrap_or(&BlockSpec::Tag(BlockTag::Latest)))?;
-        if let Some(block_number) = maybe_block_number {
-            // We cannot use `LruCache::try_get_or_insert`, because it needs &mut self, but
-            // we would need &self in the callback to construct the context.
-            if !self.block_context_cache.contains(&block_number) {
+        let block = if let Some(block_spec) = block_spec {
+            self.block_by_block_spec(block_spec)?
+        } else {
+            Some(self.blockchain.last_block()?)
+        };
+
+        if let Some(block) = block {
+            let block_header = block.header();
+            let block_number = block_header.number;
+
+            // Avoid caching of states that may contain state overrides
+            let can_cache = self.irregular_state.state_overrides().is_empty();
+
+            let contextual_state = if can_cache && self.block_state_cache.contains(&block_number) {
+                // We cannot use `LruCache::try_get_or_insert`, because it needs &mut self, but
+                // we would need &self in the callback to construct the context.
+                Cow::Borrowed(self.block_state_cache.get(&block_number).expect(
```

Comment:
I think the expect invariant doesn't hold anymore as the push is after the get in this change